### PR TITLE
Auto build test - to integration

### PR DESCRIPTION
**Pull request by lrodri29**
Here is the initial PR of the auto-build test (still needs some clean up and need to switch submodules to point to the correct branches again).

**The primary goal of this PR is to be able to perform automatic tests on our integration stack routinely. We achieve this by making use of a lambda function which is triggered on a schedule.**

NEEDS THOROUGH REVIEW

These changes should allow us to run the auto-build-test lambda function. The lambda function does the following:
1. Spin up an EC2 instance with the correct IAM role permissions to stand up the TEST stack
2. On boot a script runs which performs the following steps:
3. Install all requirements to successfully stand up a stack for the first time with the current architecture we have on integration (pre-refactor)
    - Install python3.5 and guide pip to it
    - Create dummy .aws configs
    - Install git
    - Compile Packer binary from source
    - Clone github repo and install requirements
4. Set-up log records on Cloudwatch for email service. 
5. Make vault private directory
6. Make empty aws-creds file so that cloudformation script works properly.
7. Create  keypair to attach to ec2 instances made by cloudformation.
8. Set up and configure boss-manage by sourcing a new set_vars-auto_build_test.sh bash script
9. Build AMIs
10. Create Stacks
11. Perform tests on temporary test stacks
12. Delete Stacks
13. Delete keypairs from aws

Most changes include authentication changes to critical scripts so that we can use IAM credentials rather than the aws_credentials file when running from within an instance with the correct permissions. 

I've added a manage_keypair.py script that let's us manage keypairs using the python script. 
The bash script executed on boot from the lambda function makes use of this script to create a keypair it can use to manage the instances it spins up. 

_NOTE: This PR includes the changes made to boss-manage in the vault-update PR, because they are needed to correctly configure the vault during the core stack creation, and then during the endpoint authentication. Also why I am pointing to the vault_update submodule rather than the integration branch under the boss-tools git repository._ 

TO-DO:
1. Update submodules to point back to the integration branch rather than the vault_update branch
2. Add lambda code to S3 bucket and point to it on the `set_up_auto_build_test.py` script. (This is not totally necesary to get the auto build tests to work, since we already have the lambda function on our AWS account, but important for future use).

### CODE REVIEW

File: bin/packer.py [hunk 1]
```diff
@@ -64,33 +65,37 @@ def contains(x, ys):
...
+    else:
+        with open(aws_config) as fh:
+            cred = json.load(fh)
+            session = Session(aws_access_key_id = cred["aws_access_key"],
+                            aws_secret_access_key = cred["aws_secret_key"],
+                            region_name = 'us-east-1')
+
+    client = session.client('ec2')
+    response = client.describe_images(Filters=[
+                    {"Name": "owner-id", "Values": ["099720109477"]},
```
Comment (movestill) [inline bin/packer.py]: Not your code, but where does this owner id come from?

File: cw-log-config.txt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+
```
Comment (movestill) [inline cw-log-config.txt]: What's this file used for?

File: salt_stack/salt/top.sls [hunk 3]
```diff
@@ -7,8 +7,6 @@ base:
     'vault*':
         - vault.server
         - boss-tools.bossutils
-        - scalyr
```
Comment (movestill) [inline salt_stack/salt/top.sls]: Removing the scalyr formulas will turn off logging completely for **all** stacks including production.

**Comment by movestill** [comment]
I didn't see the script that runs on boot for the ec2 instance. Any reason why you didn't download a Packer binary instead of compiling?

File: salt_stack/salt/top.sls (hunk 3 above)
Comment (lrodri29) [reply salt_stack/salt/top.sls]: Good point, I'll add this back and re-test

**Comment by lrodri29** [comment]
@movestill The script that runs on boot, will be in an S3 bucket to be used by the lambda function. I can maybe share the link to the code, but I don't think it will be necessary for the actual PR? Let me know what you think.

File: bin/packer.py (hunk 1 above)
Comment (lrodri29) [reply bin/packer.py]: Hmmm I'm not really sure...

File: cw-log-config.txt (hunk 2 above)
Comment (lrodri29) [reply cw-log-config.txt]: I use this file in the boot script (maybe should find another home for it):
```
# Set-up log records on Cloudwatch:
curl https://s3.amazonaws.com/aws-cloudwatch/downloads/latest/awslogs-agent-setup.py -O
python3 ./awslogs-agent-setup.py --region us-east-1 < ./cw-log-config.txt
```
to set the cloudwatch logs agent up, allows me to send the emails.

File: salt_stack/salt/top.sls (hunk 3 above)
Comment (movestill) [reply salt_stack/salt/top.sls]: 👍

**Comment by movestill** [comment]
Yeah I'd think it'd be helpful to include that script as part of the PR.

Also, I don't think you _have_ to use Python 3.5 on the EC2 instance just to run the CloudFormation stuff. That just happens to be the version we use on the endpoint. You're probably fine using whatever 3.x version you can install via the package manager.

**Comment by lrodri29** [comment]
@movestill Considering it will always be the same environment, I don't think the 3.5 specification is too problematic? 

And sure I'll add the script. SO you all can comment.

**Comment by movestill** [comment]
Yeah, Python 3.5 should be fine. I was worried that you were compiling it instead of just installing it since your initial PR comment said you were building Packer.

File: bin/packer.py (hunk 1 above)
Comment (derek-pryor) [reply bin/packer.py]: It is the official Ubuntu owner id, so that we get official Ubuntu images (and not from someone else)

File: cw-log-config.txt [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+
```
Comment (derek-pryor) [inline cw-log-config.txt]: In the shell script you can probably embed this as
```sh
(echo
 echo
 echo
 echo /var/log/cloud_init_output.log
 echo 1
 echo 1
 echo N) | python3 ./awslogs-agent-setup.py
```

File: lib/constants.py [hunk 5]
```diff
@@ -19,7 +19,7 @@
 # Region api is created in.  Later versions of boto3 should allow us to
 # extract this from the session variable.  Hard coding for now.
 REGION = 'us-east-1'
-INCOMING_SUBNET = "52.3.13.189/32"  # microns-bastion elastic IP
+INCOMING_SUBNET = os.environ['_BASTION_ALLOW_IP'] + "/32"  # microns-bastion elastic IP
```
Comment (derek-pryor) [inline lib/constants.py]: 1. You never added `_BASTION_ALLOW_IP` to `set_vars.sh` for developers
2. I would do `os.environ.get('_BASTION_ALLOW_IP', '52.3.13.189') + '/32'` so that existing developers don't have to change anything

File: lib/vault.py [hunk 6]
```diff
@@ -185,51 +185,42 @@ def configure(self):
...
-                client.write('/auth/aws-ec2/role/' + policy, policies = policy,
-                                                             bound_iam_role_arn = arn_prefix + policy)
-
-        # AWS Secret Backend
-        if aws_creds is None:
-            print("Vault AWS credentials file does not exist, skipping configuration of AWS secret backend")
+                client.enable_auth_backend('aws')
+            except Exception as e:
+                print("Error while enabling auth back end: " + e)
+                sys.exit(1)
```
Comment (derek-pryor) [inline lib/vault.py]: See comment I just made in the Vault PR
It looks like there are outstanding changes in the Vault PR that are not included here

File: packer/vm.packer [hunk 7]
```diff
@@ -1,8 +1,8 @@
 {
     "variables": {
         "aws_access_key": "If the default value is null, then the user variable will be required",
-        "aws_access_key": null,
-        "aws_secret_key": null,
+        "aws_access_key": "",
```
Comment (derek-pryor) [inline packer/vm.packer]: Why this change?

File: salt_stack/pillar/top.sls [hunk 8]
```diff
@@ -1,25 +1,18 @@
-base:
-  '*':
-    - scalyr
-
+base:
```
Comment (derek-pryor) [inline salt_stack/pillar/top.sls]: Doesn't look like changes we want

File: salt_stack/salt/top.sls (hunk 3 above)
Comment (derek-pryor) [reply salt_stack/salt/top.sls]: Why remove scalyr in general?

**Comment by derek-pryor** [comment]
Also, a script or documentation on how you setup the autobuild lambda would be good

File: bin/cloudformation.py [hunk 9]
```diff
@@ -93,20 +93,26 @@ def create_help(header, options):
...
 
-    if args.aws_credentials is None:
-        parser.print_usage()
-        print("Error: AWS credentials not provided and AWS_CREDENTIALS is not defined")
-        sys.exit(1)
+    if args.internal:
+        session = aws.use_iam_role()
 
+    else:
+        if args.aws_credentials is None:
```
[the remark this answers is not in the document]
Comment (lrodri29) [reply bin/cloudformation.py]: This would force us to not have AWS_CREDENTIALS as an env var or a file with credentials. Are we ok with that? I thought the internal flag would be more explicit? Can change if we want.

File: bin/manage_keypair.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+import sys
+import os
+import subprocess
+
+import alter_path
+from lib import aws
+from pathlib import Path
+
+#Define file pathname
+cur_dir = os.path.dirname(os.path.realpath(__file__))
```
[the remark this answers is not in the document]
Comment (lrodri29) [reply bin/manage_keypair.py]: Very good point. I think I made this before switching to home.... Thanks.

File: bin/manage_keypair.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+home = str(Path.home())
```
[the remark this answers is not in the document]
Comment (lrodri29) [reply bin/manage_keypair.py]: I use the home variable as a prefix to both the .pem file, and to the `/.ssh` directory

File: bin/manage_keypair.py [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+            parser.print_usage()
+            print("Error: AWS credentials not provided and AWS_CREDENTIALS is not defined")
+            sys.exit(1)
+        session = aws.create_session(args.aws_credentials)
+
+    client = session.client('ec2')
+
+    if args.action == 'create':
+        response = aws.create_keypair(session, args.keypairName)
+        subprocess.call('mkdir ' + home + '/.ssh',shell=True)
```
[the remark this answers is not in the document]
Comment (lrodri29) [reply bin/manage_keypair.py]: Oh I see. Ok I'll move to this.

File: bin/cloudformation.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (derek-pryor) [reply bin/cloudformation.py]: I don't follow. Right now (in integration) the workflow is you can specify a file on the command line or using an environmental variable and if either are given an error is printed. My comment was to change that so that instead of printing an error it just assumed that you wanted to try the boto3 defaults.

When refactored this will all be removed as we will be loading credentials using boto3 and not explicitly as we have been doing.

File: cloud_formation/lambda/auto-build-test/auto-build-test.py [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+yes | apt-get install git
+
+# Compile Packer binary from source
+wget https://releases.hashicorp.com/packer/0.12.0/packer_0.12.0_linux_amd64.zip
+unzip packer_0.12.0_linux_amd64.zip -d packer
+wait
+echo sudo mv packer /usr/bin/
+sudo mv packer /usr/bin/
+echo export PATH="/usr/bin/packer:$PATH"
+export PATH="/usr/bin/packer:$PATH"
```
[the remark this answers is not in the document]
Comment (lrodri29) [reply cloud_formation/lambda/auto-build-test/auto-build-test.py]: This is definitely needed, I remember spending quite sometime trying to get it to work without adding it to the PATH, and it would not work. I could remove and try again, but I don't think it particularly hurts to leave it there?

File: cloud_formation/lambda/auto-build-test/auto-build-test.py [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+touch config
+cd ..
+
+#Install git
+yes | apt-get install git
+
+# Compile Packer binary from source
+wget https://releases.hashicorp.com/packer/0.12.0/packer_0.12.0_linux_amd64.zip
+unzip packer_0.12.0_linux_amd64.zip -d packer
+wait
```
[the remark this answers is not in the document]
Comment (lrodri29) [reply cloud_formation/lambda/auto-build-test/auto-build-test.py]: For some reason the instance would continue to run through the boot script without waiting for the unzip command to finish.

File: cloud_formation/lambda/auto-build-test/auto-build-test.py [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+echo ' '
+echo 'echo "----------------------Init_Script----------------------'
+echo ' '
+
+#Install python3.5 and guide pip to it
+yes | add-apt-repository ppa:fkrull/deadsnakes
+yes | apt-get update
+yes | apt-get install openssh-server
+yes | ufw allow 22
+yes | apt-get install python3.5
```
[the remark this answers is not in the document]
Comment (lrodri29) [reply cloud_formation/lambda/auto-build-test/auto-build-test.py]: I can change to that `-y` use, just a style preference I suppose haha
I like having all the packages in different lines so we have a very legible way of seeing everything installed.

File: cloud_formation/lambda/auto-build-test/auto-build-test.py [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+    #  - set to shutdown the instance automatically in 60 minutes.
+    init_script = """#!/bin/bash
+echo ' '
+echo 'echo "----------------------Init_Script----------------------'
+echo ' '
+
+#Install python3.5 and guide pip to it
+yes | add-apt-repository ppa:fkrull/deadsnakes
+yes | apt-get update
+yes | apt-get install openssh-server
```
[the remark this answers is not in the document]
Comment (lrodri29) [reply cloud_formation/lambda/auto-build-test/auto-build-test.py]: Yeah I should remove this, thanks!

File: cloud_formation/lambda/auto-build-test/auto-build-test.py [hunk 17]
```diff
@@ -0,0 +1,232 @@
...
+def lambda_to_ec2(event, context):
+    """ Lambda handler taking [message] and creating a httpd instance with an echo. """
+
+    # bash script to run:
+    #  - boss-manage commands to build stack
+    #  - run tests using the instance's shell
+    #  - send all results to an S3 bucket
+    #  - send out completion SNS
+    #  - set to shutdown the instance automatically in 60 minutes.
+    init_script = """#!/bin/bash
```
[the remark this answers is not in the document]
Comment (lrodri29) [reply cloud_formation/lambda/auto-build-test/auto-build-test.py]: Oh nice! Thanks!

File: cloud_formation/lambda/auto-build-test/auto-build-test.py [hunk 18]
```diff
@@ -0,0 +1,232 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+""" Lambda to launch ec2-instances """
+import boto3
+
+REGION = 'us-east-1' # region to launch instance.
+AMI = 'ami-456b493a'
```
[the remark this answers is not in the document]
Comment (lrodri29) [reply cloud_formation/lambda/auto-build-test/auto-build-test.py]: This AMI is a very basic ubuntu 16.04 image provided by AWS. I am thinking maybe I should run the configuration steps on an instance, take snapshot of it, and use that as my new base AMI? I could maybe include the configuration up to the cloning of the repo, but no more than that. What do you think?

File: bin/cloudformation.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (lrodri29) [reply bin/cloudformation.py]: Done, way cleaner.

File: bin/packer.py [hunk 19]
```diff
@@ -164,11 +175,17 @@ def create_help(header, options):
     else:
         ami = locate_ami(credentials_config)
 
-    cmd = """{packer} build
-             {bastion} -var-file={credentials}
-             -var-file={machine} -var 'name_suffix={name}'
-             -var 'commit={commit}' -var 'force_deregister={deregister}'
-             -var 'aws_source_ami={ami}' -only={only} {packer_file}"""
+    if args.internal:
```
[the remark this answers is not in the document]
Comment (lrodri29) [reply bin/packer.py]: Good point. I don't know why I didn't use it. Thanks!

File: cloud_formation/lambda/auto-build-test/auto-build-test.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (derek-pryor) [reply cloud_formation/lambda/auto-build-test/auto-build-test.py]: I think this is a cleaner approach. Maybe a comment about the specific image used so people know what it is without searching in the AWS console.

[the remark this answers is not in the document]
Comment (lrodri29) [reply cloud_formation/lambda/auto-build-test/auto-build-test.py]: Sounds good

File: cloud_formation/lambda/auto-build-test/auto-build-test.py [hunk 20]
```diff
@@ -0,0 +1,232 @@
...
+    #  - send all results to an S3 bucket
+    #  - send out completion SNS
+    #  - set to shutdown the instance automatically in 60 minutes.
+    init_script = """#!/bin/bash
+echo ' '
+echo 'echo "----------------------Init_Script----------------------'
+echo ' '
+
+#Install python3.5 and guide pip to it
+yes | add-apt-repository ppa:fkrull/deadsnakes
```
[the remark this answers is not in the document]
Comment (lrodri29) [reply cloud_formation/lambda/auto-build-test/auto-build-test.py]: Removed

File: cloud_formation/lambda/auto-build-test/auto-build-test.py [hunk 21]
```diff
@@ -0,0 +1,232 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+""" Lambda to launch ec2-instances """
+import boto3
+
+REGION = 'us-east-1' # region to launch instance.
```
[the remark this answers is not in the document]
Comment (lrodri29) [reply cloud_formation/lambda/auto-build-test/auto-build-test.py]: leveraging the lambda env var AWS_REGIONS.

File: cloud_formation/lambda/auto-build-test/auto-build-test.py [hunk 22]
```diff
@@ -0,0 +1,232 @@
...
+python3.5 -m pip install -r requirements.txt
+python3.5 -m pip install heaviside
+
+cd salt_stack/salt/boss-tools/files/boss-tools.git
+git checkout vault_update
+cd ../../../../../
+
+# Set-up log records on Cloudwatch:
+curl https://s3.amazonaws.com/aws-cloudwatch/downloads/latest/awslogs-agent-setup.py -O
+python3 ./awslogs-agent-setup.py --region us-east-1 < ./cw-log-config.txt
```
[the remark this answers is not in the document]
Comment (lrodri29) [reply cloud_formation/lambda/auto-build-test/auto-build-test.py]: Done! Using: export EC2_REGION=`curl --silent http://169.254.169.254/latest/dynamic/instance-identity/document | jq -r .region`

File: cloud_formation/lambda/auto-build-test/auto-build-test.py [hunk 23]
```diff
@@ -0,0 +1,232 @@
...
+export SSH_KEY="~/.ssh/auto-build-keypair.pem"
+chmod 400 ~/.ssh/auto-build-keypair.pem
+# ssh-add ~/.ssh/auto-build-keypair.pem
+wait
+
+#Build AMIs
+echo " "
+echo "----------------------Building AMIs----------------------"
+echo " " 
+python3.5 ./packer.py auth vault consul endpoint cachemanager activities --name test_ami --internal
```
[the remark this answers is not in the document]
Comment (lrodri29) [reply cloud_formation/lambda/auto-build-test/auto-build-test.py]: Both of those are the same. But yeah I think these are more accurate

File: cloud_formation/lambda/auto-build-test/auto-build-test.py [hunk 24]
```diff
@@ -0,0 +1,232 @@
...
+
+echo " "
+echo "----------------------Performing Tests----------------------"
+echo " " 
+
+#Perform tests on temporary test stacks
+
+#Endpoint tests:
+echo 'Performing tests...'
+python3.5 ./bastion.py endpoint.test.boss ssh
```
[the remark this answers is not in the document]
Comment (lrodri29) [reply cloud_formation/lambda/auto-build-test/auto-build-test.py]: Oops I don't think these are up to date? I remember this not working this way... Will update. Very good catch here.

File: cloud_formation/configs/core.py [hunk 25]
```diff
@@ -125,6 +124,7 @@ def create_config(session, domain):
                                min = const.VAULT_CLUSTER_SIZE,
                                max = const.VAULT_CLUSTER_SIZE,
                                notifications = Ref("DNSSNS"),
+                               role = aws.instance_profile_arn_lookup(session, 'apl-vault'),
```
[the remark this answers is not in the document]
Comment (lrodri29) [reply cloud_formation/configs/core.py]: apl-vault is already being exported. It is not older than 4 months. 
```
...
    "InstanceProfileList": [
      {
        "InstanceProfileName": "apl-vault",
        "Path": "/"
      }
    ],
...
```

File: cloud_formation/lambda/auto-build-test/auto-build-test.py [hunk 26]
```diff
@@ -0,0 +1,232 @@
...
+        ImageId=AMI,
+        KeyName='microns-bastion20151117',
+        SecurityGroupIds=[
+            "sg-04bed83d9ddde98a1"
+        ],
+        InstanceType=INSTANCE_TYPE,
+        MinCount=1,
+        MaxCount=1,
+        IamInstanceProfile={
+            'Name': 'apl-ec2-auto-test',
```
[the remark this answers is not in the document]
Comment (lrodri29) [reply cloud_formation/lambda/auto-build-test/auto-build-test.py]: This was also previously exported [4 months ago by Sandy]
```
    ],
    "InstanceProfileList": [
      {
        "InstanceProfileName": "apl-ec2-auto-test",
        "Path": "/"
      }
    ],
```

File: lib/vault.py (hunk 6 above)
Comment (lrodri29) [reply lib/vault.py]: Will do. I'll take care of those changes in that PR, and later merge them into this branch.

File: cloud_formation/lambda/auto-build-test/auto-build-test.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (derek-pryor) [reply cloud_formation/lambda/auto-build-test/auto-build-test.py]: yea, probably meant the second to be `--name autobuild`

File: cloud_formation/lambda/auto-build-test/auto-build-test.py [hunk 27]
```diff
@@ -152,51 +160,40 @@ def lambda_to_ec2(event, context):
...
 
 # echo " "
 # echo "----------------------Cleanup environment----------------------"
 # echo " " 
 
 #Delete keypairs from aws
-python3.5 ./manage_keypair.py delete auto-build-keypair --internal
-Shutdown the instance an hour after script executes.
-shutdown -h +3600"""
+# python3.5 ./manage_keypair.py delete auto-build-keypair
```
Comment (derek-pryor) [inline cloud_formation/lambda/auto-build-test/auto-build-test.py]: Why comment these lines out?

File: bin/manage_keypair.py [hunk 28]
```diff
@@ -56,41 +51,52 @@ def create_help(header, options):
...
+        try:
+            response = aws.create_keypair(session, args.keypairName)
+            print('Protect this keypair and make sure you have access to it.')
+        except Exception as e:
+            print('Failed to create keypair due to: {}'.format(e))
+            response = False
+
+        if response:   
+            try:
+                key_file_path.touch()
```
[the remark this answers is not in the document]
Comment (lrodri29) [reply bin/manage_keypair.py]: To create the file?

File: bin/manage_keypair.py [hunk 29]
```diff
@@ -56,41 +51,52 @@ def create_help(header, options):
...
+                key_file_path.open('w').write(response['KeyMaterial'])
```
[the remark this answers is not in the document]
Comment (lrodri29) [reply bin/manage_keypair.py]: Yeah currently I chmod myself. Could make this part of the script, but it has to be done normally outside f the script anyways?

File: cloud_formation/lambda/auto-build-test/auto-build-test.py (hunk 27 above)
Comment (lrodri29) [reply cloud_formation/lambda/auto-build-test/auto-build-test.py]: I did this during my last testing session, and seems I forgot to put them back in.

File: cloud_formation/lambda/auto-build-test/auto-build-test.py [hunk 30]
```diff
@@ -152,51 +160,40 @@ def lambda_to_ec2(event, context):
 
 #Endpoint tests:
 echo 'Performing tests...'
-python3.5 ./bastion.py endpoint.test.boss ssh
-cd /srv/www/django
-python3 manage.py test
-python3 manage.py test -- -c inttest.cfg
+python3.5 ./bastion.py --ssh-key ~/.ssh/auto-build-keypair.pem endpoint.test.boss ssh-cmd "cd /srv/www/django && python3 manage.py test"# python3 manage.py test -- -c inttest.cfg
```
[the remark this answers is not in the document]
Comment (lrodri29) [reply cloud_formation/lambda/auto-build-test/auto-build-test.py]: I was getting errors, without explicitly pointing to the key. I also think it is better to explicitly see the key being used in the script's line.

File: bin/manage_keypair.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (lrodri29) [reply bin/manage_keypair.py]: Went ahead and chmod rom the script.

**Comment by lrodri29** [comment]
Minor updates are in. Thanks!

File: bin/manage_keypair.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (derek-pryor) [reply bin/manage_keypair.py]: Writing the file will also create it...

File: cloud_formation/lambda/auto-build-test/auto-build-test.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (movestill) [reply cloud_formation/lambda/auto-build-test/auto-build-test.py]: You're moving the `packer` folder to `/usr/bin`. If you did `sudo mv packer/packer /usr/bin`, you wouldn't need to change the path.

File: cloud_formation/lambda/auto-build-test/auto-build-test.py [hunk 31]
```diff
@@ -0,0 +1,227 @@
...
+mkdir .aws
+cd .aws
+touch credentials
+touch config
+cd ..
+
+#Install git
+yes | apt-get install git
+
+# Compile Packer binary from source
```
Comment (movestill) [inline cloud_formation/lambda/auto-build-test/auto-build-test.py]: This comment should just say "Download Packer".

Comment (lrodri29) [reply cloud_formation/lambda/auto-build-test/auto-build-test.py]: Done

File: cloud_formation/lambda/auto-build-test/auto-build-test.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (lrodri29) [reply cloud_formation/lambda/auto-build-test/auto-build-test.py]: Got it!
